### ISSUE TITLE
History pages not right for a number of decks

### ISSUE THREAD

**Issue by andrewwburns** [Kamalisk/arkhamdb#77]
A number of my decks seem to have lost their history, and don't show as being published (even though I just published them).

For example:

- https://arkhamdb.com/deck/edit/8535 
- https://arkhamdb.com/deck/edit/10122
- https://arkhamdb.com/deck/edit/6805
- https://arkhamdb.com/deck/edit/26225
- https://arkhamdb.com/deck/edit/26227

Some of these I'm certain used to have Histories. 

![image](https://cloud.githubusercontent.com/assets/13449428/24407401/af388682-13c2-11e7-9f9b-3ad91647d8b1.png)

Other contemporary decks still do have Histories too:

- https://arkhamdb.com/deck/edit/14510
- https://arkhamdb.com/deck/edit/14513

I can't spot any common factor between them, other than I just renamed some with Scenario numbers ("5# - Name"). Decks have been updated variously on Chrome or Safari browsers. 

What other information can I provide?

**Comment by Kamalisk** [comment]
Odd. I can take a look to see if something weird is happening with renaming. It could just be an issue loading the history rather than it being erased (hopefully that is the case)

**Comment by andrewwburns** [comment]
It does feel like that; I can't see why they would have erased. 

I took a look at how this seems to work in the page; it seems that the history is loaded as a JSON array of deck changes, separated into Changes, a decklist, and versions. 

For deck https://arkhamdb.com/deck/edit/8535, which doesn't show the history for me, I can still see the JSON, with the latest change being v1.8, on the 22-March. However, Wendy's Deck is showing at being version 4.4 to me. Ah! Intermediate changes might have been to the 'Notes' field, and not necessitated new deck list records. I'll explicitly change her deck...

Done, and I can now see v4.5, with the changes I made:
![image](https://cloud.githubusercontent.com/assets/13449428/24450452/d6f80156-1472-11e7-86cf-247825e1439c.png)

So, changes are being captured correctly, and output with the page correctly - but not rendered by JS for some reason. I'll continue to take a look at that later.

**Comment by andrewwburns** [comment]
Right, I did some more debugging last night, and I think I see what's going on...

I noticed that the variable tbody in _app.deck_history.js_, line 325 was being set to an empty jQuery object in the decks that didn't work. For the ones that do work, this is set to a **<tbody>** element of the DOM.  As it's a DOM object, I wondered if this could be due to it not being loaded yet.

I have been writing a bit of a 'History' into the 'Notes' field. They're pretty long notes, to be honest; see attached. [Notes.txt](https://github.com/Kamalisk/arkhamdb/files/881377/Notes.txt)

I wondered - what if I remove the notes? Well, I did - and the history returned. I added the notes back - and it went away. That suggests a race condition between the deck history JavaScript, and the page DOM loading.

Yup, I replicated this - I created a new deck - history shown. I added my notes - history gone. I removed again - history shown.

Therefore, I guessed that the select on _app.deck_history.js_, line 325 is running before the **tbody** exists (probably before it's been parsed).  Looking at the JavaScript block in _deckedit.html.twig_, I notice that there doesn't seem to be anything to only run **app.deck_history.init()** when the document is ready.

Can I suggest putting lines 10-11 of  inside a jQuery .ready() call? https://api.jquery.com/ready/  I think that should do it. 

There are other ways of doing this, but the jQuery syntax is very tidy.

```
$( document ).ready(function() {
  //Do Deck History Stuff Here.
});
```

It's probably a good idea to only let JavaScript that can affect the DOM run only after the DOM has finished loading - which is what the **document.onReady()** event gives you. There may be other sections of the site affected.

I don't know if there is a maximum size on the Notes field of a deck too, but that might need enforced...

(Oh, and as a side note, I think the site is fantastically useful! Thanks for your work!)

**Comment by Kamalisk** [comment]
I changed it, I have no idea if it worked, since I cant see the problem :P

**Comment by andrewwburns** [comment]
Great; no problem, I'll test it when I get a chance. It is possible that machine speed could be a factor, though I'm surprised that you couldn't replicate it with long enough notes.

**Comment by Kamalisk** [comment]
I made quite large notes, didn't seem to make much difference to me, I am perhaps closer to the server than you, do there is less latency.

**Comment by andrewwburns** [comment]
Hmm. That didn't fix it for me, so I've just tried large notes using a Lorem Ipsum generator. No problem. 

That set me wondering - could it be something specific about the Markdown I've been using? So, I've tried the Markdown, copying chunks in until I can localise if there is something specifically breaking the tab. I think I've found it. I had a section in my notes kind of like a campaign log - and I'd called it "History". The Markdown was `##History`. As soon as I add this in, the history tab fails. If I change the Markdown at all - to `##Hisstory` for example - the tab reappears...

So, not a race condition at all. 

Thinking about this, the History tab does seem to have a DIV with an ID of 'History'.

It's a bit weird, but I wonder if the problem is that the JavaScript trying to find the TBODY element is starting by selecting the History element in the Markdown? It seems a bit odd, but that would be consistent. It selects the wrong element, and then can't find the TBODY within, and the History can't be loaded into it as a result.

For now, this means a solution to me would be... don't use a title of `##History`. Anything else should be fine, and seems to be. Obviously, the best solution would be if the code getting the History section did actually get the right section...

Are you able to replicate when you use a title of `##History` in the notes?

**Comment by andrewwburns** [comment]
Ah, yeah, confirmed - with a title of History, the Markdown processor generates an `<h2>` element with an ID of "History" - which breaks the history tab.
